dev/reference/tools: add migrating data from Aurora to TiDB (#1145)
pingcap/docs: dev/reference/tools/data-migration/from-aurora.md

---
title: Data Migration from AWS Aurora MySQL to TiDB
summary: Using DM to migrate from AWS Aurora MySQL to TiDB
category: tools
---

# Data Migration from AWS Aurora MySQL to TiDB

> This document describes how to use DM to migrate from [AWS Aurora MySQL](https://aws.amazon.com/cn/rds/aurora/details/mysql-details/) to TiDB。

Reviewer: we can update links to AWS Aurora from Chinese Ver to English Ver, `https://aws.amazon.com/rds/aurora/details/mysql-details/` for this one.

(if you open the link above, it may be redirected to the Chinese Ver, don't worry because it can work for English user)

Reviewer 2: @huangxiuyan I added the link suggested by @csuzhangxc to my suggestion for this line.

> ## Step 1: Start binlog in Aurora cluster

Reviewer: ```suggestion
## Step 1: Enable binlog in Aurora cluster
```
?

This document assumes you'll migrate data from two Aurora clusters to  TiDB. The information list of Aurora clusters is as belowed. It indicates Aurora-1 has a single built-in reader endpoint. 

| Cluster | Endpoint | Port | Role |
|:-------- |:--- | :--- | :--- |
| Aurora-1 | pingcap-1.h8emfqdptyc4.us-east-2.rds.amazonaws.com | 3306 | Writer |
| Aurora-1 | pingcap-1-us-east-2a.h8emfqdptyc4.us-east-2.rds.amazonaws.com | 3306 | Reader |
| Aurora-2 | pingcap-2.h8emfqdptyc4.us-east-2.rds.amazonaws.com | 3306 | Writer |

> During the incremental backup process, you need to set the binglog format as `ROW`. Inactive binlog or incorrect binlog format will hinder data migration using DM. See [Checking items](https://www.pingcap.com/docs-cn/tools/dm/precheck/#%E6%A3%80%E6%9F%A5%E5%86%85%E5%AE%B9) for more information.

Reviewer: should link to the doc under this repo.

> During the incremental backup process, you need to set the binglog format as `ROW`. Inactive binlog or incorrect binlog format will hinder data migration using DM. See [Checking items](https://www.pingcap.com/docs-cn/tools/dm/precheck/#%E6%A3%80%E6%9F%A5%E5%86%85%E5%AE%B9) for more information.

Reviewer 2: Typo: binglog
Wrong link: why a Chinese link?

```suggestion
DM relies on the `ROW` format of binlog during the incremental replication process, so you need to set the binlog format as `ROW`. If binlog is not enabled or is incorrectly configured, DM cannot replicate data normally. For more details, see [Checking items](/dev/reference/tools/data-migration/precheck.md#checking-items).
```

> > **Warning:** 

Reviewer 2: Not serious enough to use **Warning**.

```suggestion
> **Note:** 
```

>

> > Aurora reader cannot start binlog, so it cannot be the upstream master server when using DM to migrate data.

Reviewer 2: ```suggestion
> Aurora reader cannot start binlog, so it cannot be the upstream master server when using DM to migrate data.
```
```suggestion
> Because binlog cannot be enabled in the Aurora reader, it cannot be taken as the upstream master server when you use DM to migrate data.
```

> If you need to use GTID for replication, enable GTID for Aurora cluster. 

Reviewer 2: ```suggestion
If you need to migrate data based on GTID (Global Transaction Identifier), enable GTID for the Aurora cluster.
```

> > **Warning:**

Reviewer 2: ```suggestion
> **Note:**
```

>
> GTID-based data migration requires MySQL 5.7 (Aurora 2.04.1) version or later. 

> ### Modify binlog relevent parameters in the Aurora cluster 

Reviewer 2: **Typo: relevent -> relevant**

```suggestion
### Modify binlog related parameters in the Aurora cluster 
```

> In Aurora cluster, binlog relevant parameters are in the **cluster parameter group**. See [Enable Binary Logging on the Replication Master](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Replication.MySQL.html) for more information about enabling binlog. You need to set the `binlog_format` in `ROW` when using DM for migration. 

Reviewer 2: - For a countable noun, either add an article before it or use the plural form.
- Some meaning is lost.
- The link is not accurate enough.
- set the `binlog_format` in `ROW`: `to` or `as`, but not `in`

```suggestion
In the Aurora cluster, binlog related parameters are cluster level parameters among cluster parameter groups. For more information about enabling binlog in the Aurora cluster, see [Enable Binary Logging on the Replication Master](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Replication.MySQL.html#AuroraMySQL.Replication.MySQL.EnableBinlog). You need to set the `binlog_format` to `ROW` when using DM for data migration.
```

    source-id = "mysql-replica-01"
    dm-worker = "172.16.10.72:8262"

    [[deploy]]
    source-id = "mysql-replica-02"
    dm-worker = "172.16.10.73:8262"
    ```

## Step 4: Configurate the task

This document assumes that you need to use both incremental backup and full backup of `test_table` table in `test_db` in Aurora-1 and Aurora-2 instance, and replicate to the `test_table` table of `test_db` in downstream TiDB.

Copy and edit `{ansible deploy}/conf/task.yaml.example` to generate the following config file `task.yaml`: 

```yaml
# The task name. You need to use a different name for each of the multiple tasks that run simultaneously.
name: "test"
# The full backup plus incremental backup for data replication.
task-mode: "all"
# The downstream TiDB config information.
target-database:
  host: "172.16.10.83"
  port: 4000
  user: "root"
  password: ""

# Configuration of all the upstream MySQL instances required by the current data replication task.
mysql-instances:
-
  # The ID of upstream instances or the replication group. You can refer to the configuration of `source_id` in the "inventory.ini" file or in the "dm-master.toml" file.
  source-id: "mysql-replica-01"

>   # The configuration item name of the black and white lists of the name of the database/table to be replicated, used to quote the global black and white lists configuration that is set in the global black-white-list below.

Reviewer 2: The Chinese wording is not clear enough.

```suggestion
  # The configuration item name of the black and white lists of the schema or table to be replicated, used to quote the global black and white lists configuration. For global configuration, see the `black-white-list` below.
```

  black-white-list: "global"
  # The configuration item name of mydumper, used to quote the global mydumper configuration. 
  mydumper-config-name: "global"

-
  source-id: "mysql-replica-02"
  black-white-list: "global"
  mydumper-config-name: "global"

# The global configuration of black and white lists. Each instance can quote it by the configuration item name.
black-white-list:
  global:
    do-tables:                        # The white list of the upstream table to be replicated
    - db-name: "test_db"              # The database name of the table to be replicated 
      tbl-name: "test_table"          # The name of the table to be replicated 

> # mydumper global configuration. Each instance can quote with the configuration item name. 

Reviewer 2: Keep similar description consistent.

```suggestion
# mydumper global configuration. Each instance can quote it by the configuration item name. 
```

mydumpers:
  global:

>     extra-args: "-B test_db -T test_table"  # It can only output the `test_table` of `test_db` and configure the parameter of mydumper. 

Reviewer 2: ```suggestion
    extra-args: "-B test_db -T test_table"  # Only outputs the `test_table` table of the `test_db` schema and can configure any parameter of mydumper. 
```

```

## Step 5: Start the task

> 1. Go to dmctl directory `/home/tidb/dm-ansible/resources/bin/`

Reviewer 2: ```suggestion
1. Go to the dmctl directory: `/home/tidb/dm-ansible/resources/bin/`.
```

> 2. Start dmctl using:

Reviewer 2: ```suggestion
2. Start dmctl using the following command:
```

    ```bash
    ./dmctl --master-addr 172.16.10.71:8261
    ```

> 3. Start data replication task using:

Reviewer 2: ```suggestion
3. Start data replication task using the following command:
```

    ```bash

>     # `task.yaml` is the previous edited config file 

Reviewer 2: ```suggestion
    # `task.yaml` is the previously edited configuration file.
```

    start-task ./task.yaml
    ```
    

>     - The task has been started successfully if the returned results do not report any error. 

Reviewer 2: ```suggestion
    - If the returned results do not contain any error, it indicates the task is successfully started. 
```

>     - The permission of upstream Aurora user may not be supported by TiDB if the returned results report the following errors:  

Reviewer 2: ```suggestion
    - If the returned results contain the following error information, it indicates the upstream Aurora user might have privileges unsupported by TiDB.  
```

        
        ```json
        {
            "id": 4,
            "name": "source db dump privilege chcker",
            "desc": "check dump privileges of source DB",
            "state": "fail",
            "errorMsg": "line 1 column 285 near \"LOAD FROM S3, SELECT INTO S3 ON *.* TO 'root'@'%' WITH GRANT OPTION\" ...",
            "instruction": "",
            "extra": "address of db instance - pingcap-1.h8emfqdptyc4.us-east-2.rds.amazonaws.com"
        },
        {
            "id": 5,
            "name": "source db replication privilege chcker",
            "desc": "check replication privileges of source DB",
            "state": "fail",
            "errorMsg": "line 1 column 285 near \"LOAD FROM S3, SELECT INTO S3 ON *.* TO 'root'@'%' WITH GRANT OPTION\" ...",
            "instruction": "",
            "extra": "address of db instance - pingcap-1.h8emfqdptyc4.us-east-2.rds.amazonaws.com"
        }
        ```
        

>         Using either of the following solutions, you can use `start-task` to restart the task:

Reviewer 2: ```suggestion
        To resolve this issue, use either of the following two solutions to handle it and then use the `start-task` command to restart the task:
```

>         1. Remove the unnecessary permission not supported by TiDB for Aurora user when migrating data 

Reviewer 2: ```suggestion
        1. Remove the unnecessary privileges unsupported by TiDB for the Aurora user that is used to migrate data. 
```

>         2. If Aurora user is ensured to have the permission that DM needs, then add the following configuration item in `task.yaml` config file to skip the permission precheck when start the task. 

Reviewer 2: ```suggestion
        2. If you can make sure that the Aurora user has the privileges required by DM, add the following configuration item to the `task.yaml` configuration file to skip the privileges precheck when starting the task.
```

>             ```yaml

Reviewer 2: Leave a blank line before this code block. The format is also wrong in docs-cn and you can check the wrong display at the PingCAP website [here](https://pingcap.com/docs-cn/tools/dm/from-aurora/). @huangxiuyan Please help update it too.

```suggestion

            ```yaml
```

            ignore-checking-items: ["dump_privilege", "replication_privilege"]
            ```

## Step 6: Query the task

> If you want to know whether there is a on-going data replication in DM cluster or the information about the task status, execute the following command to query:

Reviewer 2: ```suggestion
To view the on-going data replication task(s) in the DM cluster or the task status, run the following command in dmctl to query:
```

```bash
query-status
```

> > **Warning**: 

Reviewer 2: ```suggestion
> **Note:**
```

>

> > If the returned results of the command report the following error, it means the lock cannot be acquired when dumping in full backup. 

Reviewer 2: ```suggestion
> If the following error message is in the returned results of the above query command, it indicates the corresponding lock cannot be obtained during the dumping phase of the full data migration. 
```

>   ```bash
>   Couldn't acquire global lock, snapshots will not be consistent: Access denied for user 'root'@'%' (using password: YES)
>   ```

> >If the consistency between dump file and metadata is allowed not to be ensured by FTML, or the writing process in upstream can be paused, you can skip the error by adding `--no-locks` in `extra-args` under the `mydumpers`. The procedure is as follows: 

Reviewer 2: Typo: FTWL, not FTML

```suggestion
> If it is acceptable to not use FTWL to guarantee that the dump file is consistent with metadata or the upstream can pause writing data, you can skip the above error by adding the `--no-locks` argument for `extra-args` under `mydumpers`. The steps are as follows:
>
```

> > 1. Use `stop-task` to stop the paused task caused by irregular dumping 

Reviewer 2: ```suggestion
> 1. Use the `stop-task` command to stop the paused task caused by the failure of normal dumping.
```

> > 2. Update the `extra-args: "-B test_db -T test_table"` in task.yaml to `extra-args: "-B test_db -T test_table --no-locks"`

Reviewer 2: ```suggestion
> 2. In the `task.yaml` file, modify `extra-args: "-B test_db -T test_table"` to `extra-args: "-B test_db -T test_table --no-locks"`.
```

> > 3. Use `start-task` to restart the task

Reviewer 2: ```suggestion
> 3. Use the `start-task` command to restart the task.
```
